include/dlaf/common/assert.h

```diff
@@ -0,0 +1,38 @@
+//
+// Distributed Linear Algebra with Future (DLAF)
+//
+// Copyright (c) 2018-2019, ETH Zurich
+// All rights reserved.
+//
+// Please, refer to the LICENSE file in the root directory.
+// SPDX-License-Identifier: BSD-3-Clause
+//
+
+#pragma once
+
+#include <exception>
+#include <iostream>
+
+#include "dlaf/common/utils.h"
+
+#define DLAF_TEST(category, condition, ...)                                                  \
+  if (!(condition)) {                                                                        \
+    std::cerr << "[" category "-ERROR] " << __FILE__ << " : " << DLAF_SOURCE_LOCATION << ":" \
+              << __LINE__ << "\n"                                                            \
+              << dlaf::common::concat(#condition, ##__VA_ARGS__) << std::endl;               \
+    std::terminate();                                                                        \
+  }
+
+#ifndef NDEBUG
+#define DLAF_ASSERT(condition, ...) DLAF_TEST("ASSERT", condition, ##__VA_ARGS__)
+#else
+#define DLAF_ASSERT(condition, ...)
+#endif
```

Comment:
We might want to use a custom flag: e.g. `DLAF_DISABLE_CHECK` (or ASSERT) instead of NDEBUG here for two reasons:
- NDEBUG might affect projects we depend on if they use the `assert()` macro. We might want `DLAF_ASSERT` enabled even if assertions in depending projects are disabled.
- We might want to have some checks enabled in Release too.

Reply:
The idea of `DLAF_ASSERT` is to have a replacement for the `assert` basic macro, mainly allowing to have a custom message.